실시간 댓글 기능 서비스 추가

Bue-von-hon opened the pull request:
Motivation:
더 스펙타클한 발표를 위해 실시간 조회수에서 실시간 댓글로 피봇합니다.

Modification:
- Fix(build.gradle.kts): 안쓰는 의존성 삭제 및 정리
- Add(LICENSE): 라이센스 추가
- Add(README): 프로젝트 설명과 실행방법이 적힌 리드미 파일 추가
- Add(Comment): 댓글 클래스 추가
- Move(Connection, IndexController, SseController, SseEmitters): 파일이 많아져서 패키지로 묶음
- Add(PostCommentRequest): 코멘트 생성을 위한 리퀘스트 클래스 생성
- Add(CommentService): 코멘트 생성 및 커넥션을 맺기 위한 로직 추가
    - startViewingArticle: 기존과 동일하게 탭 아이디와 아티클 아이디를 인풋으로 받아, 커넥션을 생성
    - saveAndSend: 코멘트를 입력받아서 기존에 연결된 모든 커넥션에 전달받은 코멘트를 전송
- Fix(SseController): 코멘트 생성을 위한 엔드포인트 추가
- Fix(SseEmitterTest): 공통 설정을 위한 이너클래스 생성 및 코드 정리

Result:
실시간 댓글 기능을 보여주기 위한 첫번째 스텝

Bue-von-hon committed:
실시간 댓글 기능 서비스 추가

Motivation:
더 스펙타클한 발표를 위해 실시간 조회수에서 실시간 댓글로 피봇합니다.

Modification:
- Fix(build.gradle.kts): 안쓰는 의존성 삭제 및 정리
- Add(LICENSE): 라이센스 추가
- Add(README): 프로젝트 설명과 실행방법이 적힌 리드미 파일 추가
- Add(Comment): 댓글 클래스 추가
- Move(Connection, IndexController, SseController, SseEmitters): 파일이 많아져서 패키지로 묶음
- Add(PostCommentRequest): 코멘트 생성을 위한 리퀘스트 클래스 생성
- Add(CommentService): 코멘트 생성 및 커넥션을 맺기 위한 로직 추가
    - startViewingArticle: 기존과 동일하게 탭 아이디와 아티클 아이디를 인풋으로 받아, 커넥션을 생성
    - saveAndSend: 코멘트를 입력받아서 기존에 연결된 모든 커넥션에 전달받은 코멘트를 전송
- Fix(SseController): 코멘트 생성을 위한 엔드포인트 추가
- Fix(SseEmitterTest): 공통 설정을 위한 이너클래스 생성 및 코드 정리

Result:
실시간 댓글 기능을 보여주기 위한 첫번째 스텝

Bue-von-hon committed:
댓글 속성 초기화 로직 정리 및 주석 추가

수정 사항:
댓글 생성 시 좋아요수는 무조건 0부터 시작하도록 수정
public api에 대한 주석과 인스턴스 변수들에 대한 주석을 추가

Bue-von-hon commented:
@ddaaac gentle ping

ddaaac commented on src/main/java/com/example/youthcon/application/Connection.java:
```diff
@@ -1,11 +1,12 @@
 import org.springframework.web.servlet.mvc.method.annotation.SseEmitter;
 import org.springframework.web.servlet.mvc.method.annotation.SseEmitter.SseEventBuilder;
 
+import com.example.youthcon.Comment;
```
Comment 객체에 달린 synchronized 메서드들에 대해 남기는 코멘트입니다.

이미 SseEmitter 의 메서드들에 synchronized 키워드가 달려있어서 여기서까지 굳이 이 키워드를 사용할 필요가 없을 것 같아요.
다른 곳에서 사용된 synchronized 키워드에 대해서도 고민해보시면 좋을 것 같아요. 이 키워드의 남발은 꽤 큰 성능저하를 일으킬 수 있을 것 같네요.

Bue-von-hon replied:
불필요한 synchronized 키워드를 삭제했습니다!

ddaaac commented on src/main/java/com/example/youthcon/application/CommentService.java:
```diff
@@ -0,0 +1,116 @@
+
+    // 탭과 연결된 커넥션 그리고 캐시를 이용한 만료처리
+    // 타임 아웃시간이 지나도 탭을 커넥션을 들고있으면 안돼는데, 만료 시간을 설정하면 굳이 지우는 연산을 안해도된다.
+    private final Cache<String, Connection> tabIdToConnection = CacheBuilder.newBuilder()
+            .expireAfterAccess(10000L, TimeUnit.MILLISECONDS)
+            .build();
+
+    // 아티클과 연결된 커넥션, 같은 아티클을 보는 여러 탭에 이벤트 전송을 막기위함
+    // TODO(hun): 이제는 댓글이라 여러 탭에 이벤트를 전송해야..? 이전에는 푸시 알림이라 방지했지만...
+    private final Map<String, Set<Connection>> articleToConnection = new ConcurrentHashMap<>();
```
SseEmitters를 두고 새로 커넥션을 관리해주신 이유가 있을까요?

A 아티클에 새 댓글이 달리면 -> A 아티클과 연결된 Connection들에 푸쉬를 날려주면 될 것 같은데 이 동작은 이미 구현된 SseEmitters를 통해 할 수 있을 것 같아요.

Comment에 관련된 기능과 Sse에 관련된 기능을 분리하는 관점에서도 SseEmitters를 재활용하면 좋을 것 같아요.
Comment만 다루는 객체와 Sse만 다루는 객체, 그리고 이 둘을 합쳐서 관리하는 서비스가 하나 있으면 괜찮지 않을까 싶어요.

ddaaac commented on src/main/java/com/example/youthcon/application/CommentService.java:
```diff
@@ -0,0 +1,116 @@
+    private final Map<String, Set<Connection>> articleToConnection = new ConcurrentHashMap<>();
+
+    // 특정 탭에서 특정한 아티클을 보기 시작할대 호출되는 메소드
+    public synchronized Connection startViewingArticle(final String tabId, final String articleId) {
```
이 메서드가 잘 동작하는지 확인하려면 적어도 테스트 코드가 있거나, 브라우저를 통해 간단하게 코멘트를 추가하는 기능이 필요할 것 같아요. 기획안이 따로 있는 것이 아니기 때문에 단순히 프로덕션 코드만 보고 리뷰를 하기에는 정보가 부족하네요.

Bue-von-hon replied:
네넵 관련해서 커넥션연결, 댓글송수신, 타임아웃 3가지 경우에 대한 테스트를 추가했습니다

ddaaac commented on src/main/java/com/example/youthcon/application/Connection.java:
```diff
@@ -56,6 +57,15 @@ public synchronized void count(final long count) {
 
+    public synchronized void sendComment(final Comment comment) {
+        final SseEmitter.SseEventBuilder eventBuilder =
+                SseEmitter.event()
+                          .name("newComment")
+                          .data(comment)
```
클라에서 comment data를 받아서 출력해주는 단순한 로직 정도도 있으면 좋을 것 같네요.

Bue-von-hon replied:
넵 클라이언트는 물론 서버에서도 간단한 로그를 찍도록 추가했습니다.

ddaaac commented on src/main/java/com/example/youthcon/application/CommentService.java:
```diff
@@ -0,0 +1,116 @@
+
+    private void sendComment(final Comment comment, final String articleId) {
+        final Set<Connection> connections = articleToConnection.getOrDefault(articleId, new HashSet<>());
+        connections.parallelStream().forEach(connection -> connection.sendComment(comment));
```
synchronized된 메서드를 병렬로 실행해도 큰 의미가 없을 것 같아요.

Bue-von-hon replied:
삭제했습니다!

ddaaac commented on src/main/java/com/example/youthcon/application/CommentService.java:
```diff
@@ -0,0 +1,116 @@
+    }
+
+    private void updateComments(final Comment comment, final String tabId) {
+        final List<Comment> comments = articleIdToComments.computeIfAbsent(tabId, k -> new ArrayList<>());
```
여기는 articleId가 key로 쓰어야하지 않을까요?

Bue-von-hon replied:
넵 수정했습니다!

ddaaac commented on src/main/java/com/example/youthcon/application/CommentService.java:
```diff
@@ -0,0 +1,116 @@
+
+    private void sendComment(final Comment comment, final String articleId) {
+        final Set<Connection> connections = articleToConnection.getOrDefault(articleId, new HashSet<>());
+        connections.parallelStream().forEach(connection -> connection.sendComment(comment));
```
comment 저장 요청을 날린 tab을 제외하고 푸쉬를 보내주는 게 클라 처리를 좀 줄일 수 있는 방법일 것 같아요.

Bue-von-hon replied:
네넵 필터링 추가했습니다!

ddaaac reviewed (commented):
안녕하세요 태훈님 👋 

리뷰가 좀 많이 늦었네요 😭 

몇 가지 코멘트 남겼으니 확인해주세요~

Bue-von-hon committed:
댓글 저장 및 조회를 서포트하는 프론트 코드 추가

Modification:
삭제(SseEmitters.java): 사용하지 않는 코드 삭제
삭제(PostCommentRequest): 사용하지 않는 코드 삭제
수정(comment): 롬복 적용
수정(CommentService): 댓글 조회 기능 추가, 탭별이 아니라 아티클별로 댓글을 저장하게 수정, 오타 수정
수정(CommentController): Sse에서 comment로 이름 변경, SseEmitters 대신 commentService 사용하게 수정, 댓글 조회 기능 추가
수정(index.html): 댓글을 저장하고, 저장된 댓글을 불러오고, 불러 온 댓글을 리스트로 보여주는 기능 추가

Bue-von-hon committed:
댓글을 전송하고 다른 커넥션에서 댓글을 수신하는 테스트 추가

Modification:
수정(build.gradle): 불필요한 버전 정보 삭제
수정(CommentService): 아티클 아이디를 이용해 자기 자신은 제외하고 댓글을 전송하게 수정
수정(Connection): 불필요한 synchronized 키워드 삭제 및 사용하지 않는 메소드 삭제, 로그 추가
수정(CommentTest): 댓글을 전송하고, 전송한 댓글이 다른 커넥션에서 수신되는지 테스트하는 코드 추가

Bue-von-hon committed:
댓글전송수신, 타임아웃에 대한 테스트 추가

Modification:
- 수정(SseCommentTest)
    - 댓글 전송 수신 테스트: 커넥션 2개를 맺고 커넥션1에서 전송한 댓글을 커넥션2에서 수신했는지 테스트
    - 타임아웃 테스트: 커넥션을 맺고, 타임아웃까지 대기한 후 onFailure가 호출되는지 테스트

Bue-von-hon reviewed (commented):
@ddaaac @carlos-dev-yang 리뷰에 언급되었던 부분들을 전부 수정했습니다!
PTAL 🙏

Bue-von-hon committed:
캐시 만료 시간 수정 및 새로운 댓글 생성시 new 태그 추가, 유저가 탭을 닫은 경우에 대한 예외 처리 추가

Motification:
- 수정(Connection): 댓글을 전송하기 전에, 해당 커넥션과 연결된 클라이언트가 탭을 닫았으면 complte하게 수정
- 수정(CommentService): 캐시 만료 시간을 살짝 증가시켜, 실제 타임아웃 시간과 비슷하게 만듬
- 수정(index.html): 새로운 댓글이 생성되었다는 이벤트를 수신받으면, 새로운 댓글에 new태그를 5초간 달고 생성되는 로직을 추가

ddaaac commented on build.gradle.kts:
```diff
@@ -25,15 +25,18 @@ repositories {
     implementation("org.springframework.boot:spring-boot-starter-thymeleaf")
-    implementation("com.google.guava:guava:32.1.1-jre")
+    
+    implementation("com.google.guava:guava")
```
이대로 gradle 빌드를 하면 컴파일이 되지 않습니다. 확인 부탁드려요.

ddaaac commented on src/main/java/com/example/youthcon/application/CommentService.java:
```diff
@@ -0,0 +1,127 @@
+
+@Slf4j
+@Service
+public class CommentService {
```
Comment라는 도메인에 관련된 로직과 Sse라는 기술에 관한 로직이 섞여있어서 한 눈에 로직 파악이 쉽지 않은 것 같아요. 
Sse는 재활용될 확률도 높은 기술이니 만큼 Sse 관련 로직은 분리해보는 건 어떨까요? 이전 SseEmitters를 활용해볼 수 있을 것 같아요.

ddaaac commented on src/main/java/com/example/youthcon/application/CommentService.java:
```diff
@@ -0,0 +1,127 @@
+    }
+
+    // 댓글을 저장하고 연결된 커넥션들에게 댓글 생성 이벤트를 전달한다.
+    public synchronized void saveAndSend(
```
이런 synchronized도 당장은 불필요할 것 같아요. 락보다는 Sse에 집중해보는 건 어떨까요?

ddaaac commented on src/main/java/com/example/youthcon/application/CommentService.java:
```diff
@@ -0,0 +1,127 @@
+        if (oldConnection != null) {
+            try {
+                oldConnection.complete();
+            } catch (Exception e) {
```
여기에는 어떤 예외가 잡히나요??

ddaaac commented on src/test/java/com/example/youthcon/SseCommentTest.java:
```diff
@@ -0,0 +1,160 @@
+        Assertions.assertEquals("connected!", eventWrapper.receivedData.get(0));
+
+        // 연결이 타임아웃될 때까지 대기
+        sleep(10000);
```
실제 핸즈온을 할 때는 이 값을 조금 줄이는 게 좋을 것 같네요.

ddaaac commented on src/test/java/com/example/youthcon/SseCommentTest.java:
```diff
@@ -0,0 +1,160 @@
+        Assertions.assertTrue(eventWrapper.onFailureCalled);
+    }
+
+    private Response getPostResponse(String articleId1, String tabId1, Comment comment)
+            throws JSONException, IOException {
+        json.put("content", comment.getContent());
+        RequestBody requestBody = RequestBody.create(json.toString(), MediaType.parse("application/json"));
+        Request postCommentRequest = getPostCommentRequest(articleId1, tabId1, requestBody);
+        Call call = client.newCall(postCommentRequest);
+        Response postResponse = call.execute();
+        return postResponse;
+    }
+
+    private Request getPostCommentRequest(String articleId1, String tabId1, RequestBody requestBody) {
+        return new Request.Builder()
+                .url("http://localhost:" + port + "/comment?tabId=" + tabId1 + "&articleId=" + articleId1)
+                .post(requestBody)
+                .build();
+    }
+
+    private Request getConnectRequest(final String tabId, final String articleId) {
+        return new Request.Builder()
+                .url("http://localhost:" + port + "/connect?tabId=" + tabId + "&articleId=" + articleId).build();
+    }
+
+    @NotThreadSafe
+    private final class EventSourceWrapper {
```
적당히 잘 감싸주셔서 테스트 읽기가 편했습니다 👍

ddaaac reviewed (commented):
안녕하세요 태훈님 👋 

리뷰 반영 잘 해주셨네요!

간단한 코멘트들 남겼습니다~

carlos-dev-yang commented on src/main/resources/templates/index.html:
```diff
@@ -34,12 +39,14 @@
+        var receivedCommentData = JSON.parse(e.data);
+        var commentsDiv = document.getElementById('comments');
+        var commentDiv = document.createElement('div');
+        commentDiv.textContent = receivedCommentData.content; // "content"는 댓글의 내용을 담고 있는 필드로 가정합니다.
```
```javascript
var commentDiv = document.createElement('div');

// "content"는 댓글의 내용을 담고 있는 필드로 가정합니다.
commentDiv.textContent = receivedCommentData.content; 
```

이런 형태의 주석 및 개행을 추천합니다.

밑에 코드들은 주석이 다른 방식으로 달려서 코드 전체가 통일성을 띄면 좋겠습니다.

carlos-dev-yang commented on src/main/resources/templates/index.html:
```diff
@@ -83,12 +82,29 @@
+      commentDiv.textContent = receivedCommentData.content; // "content"는 댓글의 내용을 담고 있는 필드로 가정합니다.
+
+      // 새 댓글 표시를 위한 span 요소를 생성
+      var newLabel = document.createElement('span');
```
네이밍이 좀 더 이해하기 편하게 지어지면 좋을 것 같아요.
newLabel은 너무 범용적인 이름입니다. 태그를 의미하지 않고 해당 기능을 의미하면 더 좋을 것 같아요

carlos-dev-yang commented on src/main/resources/templates/index.html:
```diff
@@ -83,12 +82,29 @@
+      }, 5000);
+
+      commentsDiv.appendChild(commentDiv);
+      console.log('새 댓글을 추가했습니다.');
```
console.log를 바로 사용하는건 좋지 않습니다.
printLog, printError 등으로 로그 함수를 만들어서 호출하면 보기가 더 편할 것 같네요.

carlos-dev-yang commented on src/main/resources/templates/index.html:
```diff
@@ -115,9 +131,55 @@
+    }).then(function (response) {
+      var comments = response.data;
+      var commentsDiv = document.getElementById('comments');
+      commentsDiv.innerHTML = '';
```
여기도 다른 코드들과 형태가 같으면 좋겠습니다.

개행의 일관성이 중요합니다.
변수 선언 그룹 뒤에는 항상 개행한다던지, 반복문이나 분기분 이전에는 개행한다던지 등

carlos-dev-yang reviewed (commented):
js 코드 리뷰는 크게 할 게 없어서 보기 좋게만 구성할 수 있도록 개변 커멘트를 달았어요.

분량이 얼마 안 되고 핸즈온의 특성상 보기 좋고 바로 이해가 가는 구조로 짜면 더 좋을 것 같습니다.
일관성있게 짜는게 가장 중요하니 다시 체크 해주세요!

Bue-von-hon committed:
삭제: 발표에 사용하지 않는 코드 삭제

Bue-von-hon committed:
- 수정: 버전 명시 및 다운그레이드
    - 버전 명시: 버전을 명시하여 환경에 따라 실행되지 않는 경우를 방지
    - 버전 다운그레이드: 자바 11, 스프링 부트 2.7로 다운그레이드 하여 좀 더 제너럴하게 변경

Bue-von-hon committed:
- 수정: 인덴트 추가

Bue-von-hon committed:
수정: 주석 위치 변경

Bue-von-hon committed:
수정: 버전 명시

Bue-von-hon committed:
수정: 새 댓글 표시를 위한 span 추가, 에러 메시지를 한글로 수정

Bue-von-hon committed:
수정: 현실적인 재요청 시간으로 수정

Bue-von-hon committed:
수정: 현실적인 타임아웃 시간으로 수정

Bue-von-hon committed:
수정: var 대신 let으로 변경, newLabel 대신 newCommentIndicator로 수정

Bue-von-hon committed:
수정: 탭 아이디 대신 유저 이름으로 수정, 유저 이름을 직접 작성할 수 있게 수정

Bue-von-hon committed:
수정: 콘솔 로그 대신 printLog 사용하게 변경

Bue-von-hon committed:
- 수정: 미리 제공될 코드와 핸즈온 당일 직접 작성할 코드를 패키지 2개로 구분
    - preparation: 미리 제공될 코드
    - handson: 당일 함께 작성하며 완성시킬 코드

Bue-von-hon committed:
- 수정: 롬복 의존성 제거, 발표 당일 원활한 진행을 위해 롬복 삭제

Bue-von-hon committed:
- 삭제: 사용하지 않는 코드 삭제

Bue-von-hon committed:
- 수정: 각각의 자료구조에 대한 클래스를 작성하여 서비스 코드 간소화

Bue-von-hon committed:
- 수정: deleteConnectionFromArticleToConnection 메소드에서 2개의 매니저 클래스에 의존하는 부분을 끊어 내고, 개별 매니저 클래스에 로직을 내재화하여 서비스 코드를 간소화

Bue-von-hon committed:
삭제: 사용하지 않는 코드 삭제

Bue-von-hon committed:
삭제: 인텐트 수정

Bue-von-hon committed:
- 추가(ConnectionManager): 탭아이디-커넥션, 아티클-커넥션 2가지 커넥션을 관리하는 클래스 추가
    - 커넥션 관련 작업은 모두 이 클래스에서 수행하도록 변경
    - 기존의 콜백을 등록하는 코드를 이 클래스에 내재화하여 서비스 코드 간소화

Bue-von-hon committed:
- 수정: 불필요한 메소드 추출을 삭제

Bue-von-hon committed:
- 삭제: 사용하지 않는 코드 삭제

Bue-von-hon committed:
- 수정: 코드 위치 변경

Bue-von-hon commented:
@ddaaac @carlos-dev-yang 리뷰남겨주신 부분들을 해결했습니다! PTAL 🙏

Bue-von-hon committed:
수정: 2차 리허설 내용 반용
  - 커넥션 매니저 3인방을 삭제하고, sse를 직접 사용하도록 수정
  - 디비에 댓글을 저장하는 대용으로 만든 아티클 투 코멘트 클래스 삭제
  - 이미 유저가 있다고 가정하고, 탭 아이디 관련 코드를 삭제

Bue-von-hon committed:
수정: 인덱스 파일 레이아웃 수정
